tests/nativeapp/factories.py

```diff
@@ -0,0 +1,239 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+import os
+from dataclasses import dataclass
+from pathlib import Path
+
+import factory
+import yaml
+
+"""
+Factories to configure project definitions and write PDF and other files on disk for testing
+
+
+"""
+
+
+class FactoryNoEmptyDict(factory.DictFactory):
+    """
+    Dict Factory that returns None if empty, instead of {}
+    """
+
+    @classmethod
+    def _create(cls, *args, **kwargs):
+        if len(kwargs) == 0:
+            return None
+        return cls._build(*args, **kwargs)
+
+
+class PackageFactory(FactoryNoEmptyDict):
+    """
+    Package V1.* Factory for creating dict representing package in PDF.
+
+    Usage: PackageFactory(name="pkg_name", role="package_role")
+
+    Package model has no required fields, therefore this is a simple NoEmptyDict factory.
+    """
+
+    pass
+
+
+class ApplicationFactory(FactoryNoEmptyDict):
+    """
+    Application V1.* Factory for creating dict representing application in PDF.
+
+    Usage: ApplicationFactory(name="my_app", role="app_role")
+
+    Application model has no required fields, therefore this is a simple NoEmptyDict factory.
+    """
+
+    pass
+
+
+class ArtifactFactory(factory.ListFactory):
+    """
+    List Factory for creating an artifact list.
+
+    Usage:
+        - ArtifactFactory(["setup.sql", "README.md"])
+        - ArtifactFactory([
+                {"src": "app/*", "dest": "./", "processors":["processor1"]},
+                {"src": "setup.sql", "dest": "setup.sql"}
+        ])
+    """
+
+    pass
+
+
+class NativeAppFactory(factory.DictFactory):
+    """
+    Factory for preparing native app dict.
+
+    Usage:
+        Create a native app with a faker-generated name and an empty artifacts list:
+        - NativeAppFactory()
+
+        Create a native app with the given name and artifacts list:
+        - NativeAppFactory(name="my_app", artifacts=[{"src": "app/*", "dest": "./"}])
+
+        Creates a native app dict with package role set to "pkg_role":
+        - NativeAppFactory(name="my_app", artifacts=["setup.sql", "README.md"], package__role="pkg_role")
+    """
+
+    name = factory.Faker("word")
+    artifacts = factory.List([], list_factory=ArtifactFactory)
+    package = factory.SubFactory(PackageFactory)
+    application = factory.SubFactory(ApplicationFactory)
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
+        if kwargs["package"] is None:
+            kwargs.pop("package")
+        if kwargs["application"] is None:
+            kwargs.pop("application")
+        return cls._build(model_class, *args, **kwargs)
+
+
+@dataclass
+class PdfFactoryResult:
+    def __init__(self, yml: dict, path: Path):
+        self.yml = yml
+        self.path = path
```

Comment:
since this is a dataclass
```suggestion
    yml: dict
    path: Path
```
the `__init__` will be auto-generated

Reply:
done!